ggml-qnn.cpp

```diff
@@ -0,0 +1,3425 @@
+//
+//  implementation of QNN backend for GGML
+//
+// =================================================================================================
+static bool ggml_qnn_can_handle_op(ggml_backend_qnn_context * ctx,
+                                   const struct ggml_tensor * tensor,
+                                   bool b_dump_tensor_info) {
+    if (ggml_is_empty(tensor) || tensor->op == GGML_OP_RESHAPE ||
+        tensor->op == GGML_OP_TRANSPOSE || tensor->op == GGML_OP_VIEW ||
+        tensor->op == GGML_OP_PERMUTE || tensor->op == GGML_OP_NONE) {
+        return false;
+    }
+
+    const struct ggml_tensor * src0 = tensor->src[0];
+    const struct ggml_tensor * src1 = tensor->src[1];
+    if (nullptr == src0 || nullptr == src1) {
+        return false;
+    }
+
+    const int64_t ne00 = src0->ne[0];
+    const int64_t ne01 = src0->ne[1];
+    const int64_t ne10 = src1->ne[0];
+    const int64_t ne11 = src1->ne[1];
+    // make qnn_get_ggml_tensor_rank and QNN SDK happy
+    if (ne00 <= 1 || ne01 <= 1 || ne10 <= 1 || ne11 <= 1) {
+        return false;
+    }
+
+    // TODO: support other GGML OPs using QNN API
+    // a GENERAL approach could fix this problem in a standalone PR of refine ggml backend
+    // subsystem for hybrid inference between CPU&GPU / CPU&NPU easily(less the 100 LoC and no
+    // side-effect to the existing codes) for ANY ggml backends which the backend's
+    // ggml_backend_xxx_buffer_is_host return true. this approach could be found at:
+    // https://github.com/ggerganov/llama.cpp/pull/7641
+    bool supported_op = false;
+    supported_op = (tensor->op == GGML_OP_ADD);
+    supported_op = ((tensor->op == GGML_OP_ADD) || (tensor->op == GGML_OP_MUL_MAT));
+    if (!supported_op) {
+        return false;
+    }
+
+    //TODO: support other quantized data type
+    if (ggml_is_quantized(src0->type)) {
+        if (src0->type != GGML_TYPE_Q8_0 && src0->type != GGML_TYPE_Q4_0) {
+            return false;
+        }
+    }
+
+    if (tensor->op == GGML_OP_MUL_MAT) {
+        if (ne00 <= 32 || ne01 <= 32 || ne10 <= 32 || ne11 <= 32) {
+            //comment it for make UT of mul_mat with QNN RPC happy
+            //return false;
+        }
+    }
+
+    return true;
+}
+
+//TODO: this function can be removed later because there are duplicated codes with ggml_qnn_mul_mat
+//      keep it for illustrate how to implement a specified GGMPL OP using QNN API + QNN RPC
+static void ggml_qnn_add(ggml_backend_qnn_context * ctx, const ggml_tensor * src0,
+                         const ggml_tensor * src1, ggml_tensor * dst) {
+    Qnn_ErrorHandle_t error             = QNN_SUCCESS;
+    bool              graph_initialized = false;
+    qnn_instance *    instance          = nullptr;
+    std::string       graph_name        = "ggml_op_qnn_add";
+    Qnn_GraphHandle_t graph_handle      = nullptr;
+    Qnn_Tensor_t *    tensor_0          = nullptr;
```

Comment:
Created a PR on your fork, to simpilify the binding from `Qnn_Tensor_t` to `ggml_tensor`, please have look if have time: https://github.com/zhouwg/llama.cpp/pull/2